File: algorithms/src/main/java/org/baichuan/sample/algorithms/leetcode/middle/Operations.java
package org.baichuan.sample.algorithms.leetcode.middle;

/**
 * @author: tk (rivers.boat.snow@gmail.com)
 * @date: 2022/1/10
 * https://leetcode-cn.com/problems/operations-lcci/
 * 面试题 16.09. 运算
 */
public class Operations {
    static final int NEGATIVE_UNIT = -1;
    static final int POSITIVE_UNIT = 1;

    public static void main(String[] args) {
        System.out.println(new Operations().minus(2, 1));
        System.out.println(new Operations().multiply(0, -5));
        System.out.println(new Operations().divide(2, 2));
    }

    public Operations() {

    }

    public int minus(int a, int b) {
        if (b < 0 && a > 0 || a < 0 && b > 0) {
            return a + b;
        }

        int tmp;

        if (b > 0) {
            tmp = NEGATIVE_UNIT;
            while (b + tmp > 0) {
                tmp += NEGATIVE_UNIT;
            }
        } else {
            tmp = POSITIVE_UNIT;
            while (b + tmp < 0) {
                tmp += POSITIVE_UNIT;
            }
        }
        return a + tmp;
    }

    public int multiply(int a, int b) {
        if (a == 0 || b == 0) {
            return 0;
        }
        int originalA = a;
        int originalB = b;
        //乘积为正数
        if (b > 0 && a > 0) {
            while (b > 1) {
                a += originalA;
                b += NEGATIVE_UNIT;
            }
            return a;
        }
        if (a < 0 && b < 0) {
            //求出a对应的正数值
            int negativeA = getPositive(a);
            originalA = negativeA;
            while (b < -1) {
                negativeA += originalA;
                b += POSITIVE_UNIT;
            }
            return negativeA;
        }

        //乘积为负数
        if (a < 0) {
            while (b > 1) {
                a += originalA;
                b += NEGATIVE_UNIT;
            }
            return a;
        } else {
            while (a > 1) {
                b += originalB;
                a += NEGATIVE_UNIT;
            }
            return b;
        }
    }

    public int divide(int a, int b) {
        int result = 0;
        if (a > 0 && b > 0) {
            while (a >= b) {
                b += b;
                result++;
            }
        } else if (a < 0 && b < 0) {
            while (a <= b) {
                b += b;
                result++;
            }
        } else if (a > 0) {
            int positiveB = getPositive(b);
            while (a >= positiveB) {
                positiveB += positiveB;
                result += NEGATIVE_UNIT;
            }
        } else {
            int negativeB = getNegative(b);
            while (a <= negativeB) {
                negativeB += negativeB;
                result += NEGATIVE_UNIT;
            }
        }

        return result;
    }

    /**
     * 获取x对应的负数
     * x>0
     *
     * @param x
     * @return -x
     */
    private int getNegative(int x) {
        int result = 0;
        while (x > 0) {
            x += NEGATIVE_UNIT;
            result += NEGATIVE_UNIT;
        }
        return result;
    }

    /**
     * 获取x对应的正数
     * x<0
     *
     * @param x
     * @return -x
     */
    public int getPositive(int x) {
        int result = 0;
        while (x < 0) {
            x += POSITIVE_UNIT;
            result += POSITIVE_UNIT;
        }
        return result;
    }
}

File: algorithms/src/main/kotlin/org/baichuan/sample/algorithms/leetcode/middle/Operations.kt
package org.baichuan.sample.algorithms.leetcode.middle

/**
 * @author: tk (rivers.boat.snow@gmail.com)
 * @date: 2022/1/11
 */
class Operations {
    var negatives = LongArray(32)
    var positives = LongArray(32)
    var cacheA = LongArray(32)
    var cacheB = LongArray(32)
    val NEGATIVE_UNIT = -1

    fun Operations() {
        negatives[0] = NEGATIVE_UNIT.toLong()
        positives[0] = 1
        for (i in 1..31) {
            negatives[i] = negatives[i + NEGATIVE_UNIT] + negatives[i + NEGATIVE_UNIT]
            positives[i] = positives[i + NEGATIVE_UNIT] + positives[i + NEGATIVE_UNIT]
        }
    }

    fun minus(a: Int, b: Int): Int {
        var aVar = a
        var bVar = b
        if (aVar == bVar) {
            return 0
        }
        var index = 31
        while (index >= 0) {
            if (bVar > 0) {
                if (bVar >= positives[index]) {
                    bVar += (negatives[index]).toInt()
                    aVar += (negatives[index]).toInt()
                } else {
                    index += NEGATIVE_UNIT
                }
            } else {
                if (bVar <= negatives[index]) {
                    bVar += (positives[index]).toInt()
                    aVar += (positives[index]).toInt()
                } else {
                    index += NEGATIVE_UNIT
                }
            }
        }
        return aVar
    }

    fun multiply(a: Int, b: Int): Int {
        var aVar = a
        var bVar = b
        if (aVar == 0 || bVar == 0) return 0
        if (aVar == 1) return bVar
        if (bVar == 1) return aVar
        if (aVar == NEGATIVE_UNIT) return minus(0, bVar)
        if (bVar == NEGATIVE_UNIT) return minus(0, aVar)
        val aSign = aVar > 0
        //最终结果的正负，true：正 false：负
        val resultSign = aVar > 0 && bVar > 0 || aVar < 0 && bVar < 0
        /**
         * 转负数是为了简化代码
         * 为什么不转正数？因为-2147483648无法转成有效的4字节正数
         */
        if (aVar > 0) {
            aVar = minus(0, aVar)
        }
        if (bVar > 0) {
            bVar = minus(0, bVar)
        }
        cacheA[0] = aVar.toLong()
        for (i in 1..30) {
            cacheA[i] = cacheA[i + NEGATIVE_UNIT] + cacheA[i + NEGATIVE_UNIT]
        }
        var result = 0
        var index = 31
        while (index >= 0) {
            if (bVar <= negatives[index]) {
                bVar += (positives[index]).toInt()
                result += (cacheA[index]).toInt()
            }
            index += NEGATIVE_UNIT
        }

        //原始结果符号，用来判断计算结果是否溢出
        val originalResultSign = result > 0
        //确保溢出的情况不会影响结果的正负
        if (originalResultSign != aSign) {
            result = minus(0, result)
        }

        //确定最终结果的正负
        if (resultSign && result < 0 || !resultSign && result > 0) {
            result = minus(0, result)
        }
        return result
    }

    fun divide(a: Int, b: Int): Int {
        var aVar = a
        var bVar = b
        if (aVar == 0 || aVar > 0 && aVar < bVar || aVar < 0 && aVar > bVar) return 0
        var result = 0
        val resultSign = aVar > 0 && bVar > 0 || aVar < 0 && bVar < 0
        if (aVar > 0) {
            aVar = minus(0, aVar)
        }
        if (bVar > 0) {
            bVar = minus(0, bVar)
        }

        //cacheB[0] = b* 2^0
        cacheB[0] = bVar.toLong()
        for (i in 1..31) {
            cacheB[i] = cacheB[i + NEGATIVE_UNIT] + cacheB[i + NEGATIVE_UNIT]
        }
        var index = 31
        var preDividedB: Long = 0
        while (index >= 0) {
            if (aVar <= cacheB[index] + preDividedB) {
                preDividedB += cacheB[index]
                result += (positives[index]).toInt()
            }
            index += NEGATIVE_UNIT
        }

        //确定最终结果的正负
        if (resultSign && result < 0 || !resultSign && result > 0) {
            result = minus(0, result)
        }
        return result
    }
}